### ISSUE TITLE
sitewide, CMS - 508-defect-2 [FOCUS MGMT]: Ensure focus moves when same page links are activated (IE11)

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#15244]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Jennifer

## User Story or Problem Statement

As a screen reader and/or keyboard user, I expect focus to move to the target location when I activate a same page link.

## Details

The VA 508 Office found one issue during their review. It is specific to IE11. The focus for same page links **must** move to the target location. There is a known IE bug where this is broken. 

## Acceptance Criteria

- [ ] Focus moves to the target `id` when a same page link is activated
- [ ] Focus outline is not rendered if the target is not interactive

## Environment

* Operating System: Windows, latest
* Browser: IE11
* Screenreading device: JAWS, but the behavior is evident without the screen reader
* Server destination: production

## Steps to Recreate

1. Enter `https://www.va.gov/decision-reviews/` in browser
2. Navigate to the On this page component
3. Using only the keyboard, tab to the second bulleted link
4. Activate the link
5. Observe the new location displayed in the browser viewport
6. Tab again
7. Verify that focus remained in the On this page bulleted list of links 

## Solution (if known)

**Please note:** Another solution will be provided in a couple of hours. I believe @1Copenut  has some JavaScript that is typically used.

### Current code

```html
<!-- On this page component -->
<h2 class="vads-u-margin-bottom--2 vads-u-font-size--lg" id="on-this-page">On this page</h2>
<ul>
          <li>
            <a href="#request-a-decision-review-or-appeal">
              Request a decision review or appeal
            </a>
          </li>
          <li>
            <a href="#manage-your-decision-reviews-and-appeals">
              Manage your decision reviews and appeals
            </a>
          </li>
          <li>
            <a href="#more-information-and-resources">
              More information and resources
            </a>
          </li>
        </ul>

<!-- target example -->
  <h2 id="request-a-decision-review-or-appeal" class="hub-page-link-list__title">
    Request a decision review or appeal
  </h2>
```

### Recommendation

Two options to consider:

1. https://stash.uconn.edu/users/jmr06005/repos/wordpress-josh/browse/wp-content/themes/twentysixteen/js/skip-link-focus-fix.js?at=578e32cb3718379cded8971a7d25cebf1d77d220

2. From the Jim Thatcher link in WCAG or vendor guidance section (shown below): https://www.jimthatcher.com/news-070607.htm

```html
<!-- On this page component -->
<h2 class="vads-u-margin-bottom--2 vads-u-font-size--lg" id="on-this-page">On this page</h2>
<ul>
          <li>
            <a href="#request-a-decision-review-or-appeal">
              Request a decision review or appeal
            </a>
          </li>
          <li>
            <a href="#manage-your-decision-reviews-and-appeals">
              Manage your decision reviews and appeals
            </a>
          </li>
          <li>
            <a href="#more-information-and-resources">
              More information and resources
            </a>
          </li>
        </ul>

<!-- target example -->
  <h2 class="hub-page-link-list__title">
    <span style="position:absolute;">
      <a name="request-a-decision-review-or-appeal" id="request-a-decision-review-or-appeal">&nbsp;</a>
    </span>
    Request a decision review or appeal
  </h2>
```

## WCAG or Vendor Guidance (optional)

* [Keyboard focus **MUST** follow the action (e.g. the focus goes to a dialog when the dialog is activated, and when the dialog is dismissed, the focus returns to the button that activated the dialog). WCAG 2.4.3](https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html)

* [Human-readable translation of the W3C WCAG 2.4.3 success criteria](https://www.wuhcag.com/focus-order/#:~:text=Your%20users%20need%20to%20find,focus%20order'%20of%20your%20website.)

* [In-page links and the IE bug, Jim Thatcher](https://www.jimthatcher.com/news-070607.htm)

* [Keyboard Navigation and Internet Explorer, Gez Lemon](https://juicystudio.com/article/ie-keyboard-navigation.php)

* [In-Page Links and Input Focus [Again], AccessibilityCulture.org](http://accessibleculture.org/articles/2010/05/in-page-links/)

**Comment by joshkimux** [comment]
I've just tested and confirmed that this issue still needs to be fixed. @brianalloyd could we prioritize this asap? This was flagged by the VA 508 office half a year ago 🙇

**Comment by brianalloyd** [comment]
@joshkimux we will prioritize and place in this sprint for action.  Thanks for following up.

**Comment by ncksllvn** [comment]
This would be sort of a medium risk change btw. Think I would rate this a 5.

**Comment by kelsonic** [comment]
@joshkimux is this still a ticket we should do? e.g. putting anchor tags within our heading tags for TOC same-page linking

**Comment by joshkimux** [comment]
Yes, but I recommend following the articles linked closely to ensure it's an anchor tag without an `href`

**Comment by Rleahy22** [comment]
https://user-images.githubusercontent.com/3144003/132895144-b62e6b22-016a-4627-8adc-1988fbc58c2e.mov

@joshkimux I've got this change locally and added `tabindex="-1"` to the `a` tags so they're not focusable when tabbing generally, but it does look this this now when you tab to and click the in page link. Is that expected/what we want?

**Comment by joshkimux** [comment]
@Rleahy22 Nice! Any chance I can test this myself using SOCKS?

**Comment by Rleahy22** [comment]
I can clean this up and open a pr if you want, but first I wanted to make sure that it's okay to have that focus outline next to the header.

**Comment by joshkimux** [comment]
@Rleahy22 We wouldn't want to have the focus outline to be visible if possible 🙏

**Comment by Rleahy22** [comment]
Fixed. I have a pr up now with an updated screen recording, and it should create a branch deployment.

**Comment by Rleahy22** [comment]
For validation, this should not affect use on any browser other than IE11. Changes should go out with the Tuesday (9/14) deployment.

**Comment by timwright12** [comment]
@joshkimux @Rleahy22  Just wanted to poke my head in here as I saw this PR get merged. It looks to me like we're violating two WCAG rules with this fix:

- [Link Purpose](https://www.w3.org/WAI/WCAG21/quickref/#link-purpose-in-context)
- [Focus Visible](https://www.w3.org/WAI/WCAG21/quickref/#focus-visible)

Link purpose fill trigger an issue on an empty link and focus visible, while it won't trigger an automated error it will cause visible focus to be lost on the link.

**Comment by Rleahy22** [comment]
Did those issues not exist prior to this change as well? What would we like these links to do in that case?

**Comment by timwright12** [comment]
@Rleahy22 these would be new issues that care coming up with the PR. In dealing with the IE11 issue, it's standard to add some JS to fix it. This is how WordPress deals with it: https://github.com/Automattic/_s/issues/1206 -- basically a small amount of JS that binds to the click of a skip link, adds `tabindex="-1"` to the target element, then sets focus to that element. There's also a blur event that removes the `tabindex="-1"` so it doesn't hang around afterwards.

The PR isn't deployed to prod yet and we run a11y tests each morning, so this should throw and error in [#-daily-accessibility-scan](https://dsva.slack.com/archives/C01RAS1KAQK) tomorrow morning. Ideally we should try to fix it up before that.

**Comment by joshkimux** [comment]
Thanks for linking to how WordPress handles it! 🙏

**Comment by Rleahy22** [comment]
@timwright12 @joshkimux since we use `h2` tags for the in page link anchors we can't use `.focus()` on them so I have 2 proposals for how we can use `a` tags.

1) Wrap the `h2` in an `a` tag with link styles reset. This makes it so the link has text unlike in the initial solution which used empty links. This solves our issue, but does make the Headings clickable, and the resulting scroll make catch the user off guard:
<img width="1784" alt="Screen Shot 2021-09-15 at 3 39 35 PM" src="https://user-images.githubusercontent.com/3144003/133501135-4c636a3b-c3bb-48ea-9730-b0024f3eec74.png">

2) Use a link icon to denote the anchor:
<img width="1785" alt="Screen Shot 2021-09-15 at 3 42 04 PM" src="https://user-images.githubusercontent.com/3144003/133501192-04da8ec7-0ea9-4106-91a4-4e7fb84b5f72.png">

Thoughts?

**Comment by joshkimux** [comment]
@Rleahy22 By any chance, did you test using `&nbsp;` instead of `tabindex="-1"` as listed in the original ticket? 

```html
<h2 class="hub-page-link-list__title">
    <span style="position:absolute;">
      <a name="request-a-decision-review-or-appeal" id="request-a-decision-review-or-appeal">&nbsp;</a>
    </span>
    Request a decision review or appeal
  </h2>
```

It's [solution A in the original article cited by Jen](https://www.jimthatcher.com/news-070607.htm) as opposed to solution B which is an empty link

**Comment by Rleahy22** [comment]
Ah I did, but it affected display. I could make the same style changes as I did in solution 1 though. Thanks @joshkimux

**Comment by Rleahy22** [comment]
@joshkimux it looks like the non breaking space doesn't meet our requirements either. I get this error with that solution as well.
<img width="897" alt="Screen Shot 2021-09-22 at 4 10 27 PM" src="https://user-images.githubusercontent.com/3144003/134414806-806c1f01-72a8-4956-85ac-14d73a296804.png">

**Comment by joshkimux** [comment]
Let me test this separately! I'll get back to you before the end of the week 😄

**Comment by joshkimux** [comment]
@Rleahy22 Alright, getting back to you now! I should have read the earlier comments in more detail 🙇 

> since we use h2 tags for the in page link anchors we can't use .focus()
`.focus()` will work for `h2` provided you set `tabindex="-1"` as an attribute. [Here's a codepen demonstrating the steps to do so!](https://codepen.io/2kool2/pen/bxdzEJ)

The solution Tim suggested should work! The codepen I attached explains how to set focus to a heading e.g. `h2` and a content block e.g. `main` 😄

**Comment by Rleahy22** [comment]
Alright, third times the charm! @joshkimux this should go out with the deployment today.

**Comment by brianalloyd** [comment]
@joshkimux could validate the work here and let us know what additional work if any is needed.  Appreciate @Rleahy22 for taking the lead and moving this forward, awesome work Ryan.

**Comment by Rleahy22** [comment]
@joshkimux good catch. The fix for that should go out in today's deploy

**Comment by brianalloyd** [comment]
@joshkimux this should be ready to validate and close.  Thanks to @Rleahy22 for the solid work on resolving this issue.

**Comment by joshkimux** [comment]
Validated. @Rleahy22 is on a roll 🔥
